packages/adminrouter/extra/src/README.md

```diff
@@ -870,3 +1112,842 @@ expressions/statements:
+    # DNS TTL override in `resolver` config option -> 5s * 1.5 = 7.5s
+    cache_poll_period = 8
+    cache_expiration = cache_poll_period - 1
+    cache_first_poll = 1
+
+    mocker.send_command(
+        endpoint_id='http://127.0.0.3:5050',
+        func_name='enable_extra_agent',
+        )
+
+    ar = nginx_class(
+        cache_first_poll_delay=cache_first_poll,
+        cache_poll_period=cache_poll_period,
+        cache_expiration=cache_expiration,
+        upstream_mesos="http://leader.mesos:5050",
+        )
+
+    with GuardedSubprocess(ar):
+        # Force cache update by issuing a request
+        ping_mesos_agent(
+            ar,
+            valid_user_header,
+            expect_status=404,
+            agent_id=AGENT_EXTRA_ID)
+
+        # Now, let's change DNS entry to point to other Mesos master
+        dns_server_mock.set_dns_entry(
+            'leader.mesos.', ip='127.0.0.3', ttl=self.LONG_TTL)
+
+        # Wait for cache to expire and let DNS entry be re-resolved
+        # during the refresh
+        time.sleep(cache_poll_period + cache_first_poll + 1)
+
+        # Make sure that cache now used the right upstream
+        ping_mesos_agent(
+            ar,
+            valid_user_header,
+            expect_status=200,
+            endpoint_id='http://127.0.0.4:15003',
+            agent_id=AGENT_EXTRA_ID)
+```
+### Advanced usage
+There are no general guidelines on how to add advanced test cases to the test
+harness, as each case is different. Sometimes one needs to add support for a
+new side effect to Mocker endpoints, sometimes it's adding a custom benchmarking
+tool to the harness in order to stress AR in certain way, sometimes it is
+just a simple test checking >40 endpoints, every one of them in a slighly
+different way, etc...
+
+In such cases it is important to keep in mind few simple rules, with which
+the whole test harness was designed and implemented:
+* *try to avoid creating custom/very specialized DSL*
+
+The more abstractions there are, the more AR test-harness-specific
+functions/shortcuts there are, the more difficult it is for people to start
+lerning it effectivelly and/or not to try to invent their own indirection
+layers. Try to keep things simple!
+
+* *try to keep the code DRY*
+
+Admin Router tests can be very repetitive. As trivial and simple it sounds, the
+more tests there are the more difficult is to refactor code and to keep it
+tidy. This increases the barier of entry for new contributors and makes them
+feel that `it is OK to do it` - it is not. If you wrote the same tests more
+than three times, then maybe it is time to implement another generic tests in
+YAML config file and start operating only on the test's input data instead.
+
+* *check for a-prior art*
+
+It is possible that the solution that you are looking for has already been
+found by somebody else. Try to search for tests that check similar behaviour or
+feature.  If you need to implement a new side effect/functionality - try
```

Comment:
`s/functionality/new functionality` - else it could be that this means "a new functionality" which is not grammatically correct.